source/lambda_functionbsm_thea.py

```diff
@@ -0,0 +1,187 @@
+the new JSON newline file, perform data transformation, and upsert the new data
+records to the corresponding Socrata data set on data.transportation.gov.
+
+Data transformation includes flattening of the data structure, which is
+required for the data to work with the Socrata backend on data.transportation.gov.
+Renaming of certain fields is done to achieve consistency across data sets. No
+unit conversion will be done to the data, though occasionally, additional fields
+will be added to enhance usage of the data set in Socrata (e.g. geolocation fields
+for plotting). Any added fields will be listed in the description of each Socrata
+data set.
+
+A separate lambda function using the lambda handler in `lambda_function_cleanup.py`
+will be used to keep the data set at a manageable size by removing oldest records
+from the Socrata data set.
+
+Requires:
+- Uploading to AWS requires packaging the source code with the Python extensions
+into a single zip file. This code requires the following Python libraries to be
+packaged with the source code to work on AWS: requests, sodapy. The file
+`lambda_to_socrata_util.py` that is in this repo should also be included.
+- The user must also set the following Environment Variables in their lambda
+function: SOCRATA_USERNAME, SOCRATA_PASSWORD, SOCRATA_API_KEY, SOCRATA_DATASET_ID.
+
+'''
+import boto3
+import dateutil.parser
+import json
+import logging
+import os
+import random
+import requests
+from requests.auth import HTTPBasicAuth
+requests.packages.urllib3.disable_warnings()
+from sodapy import Socrata
+import time
+
+import lambda_to_socrata_util
+
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+SOCRATA_USERNAME = os.environ['SOCRATA_USERNAME']
+SOCRATA_PASSWORD = os.environ['SOCRATA_PASSWORD']
+SOCRATA_API_KEY = os.environ['SOCRATA_API_KEY']
+SOCRATA_DATASET_ID = os.environ['SOCRATA_DATASET_ID']
+
+
+def process_bsm(raw_rec):
+    '''
+    Reads each raw BSM data record from Tampa CV Pilot and performs data transformation,
+    including:
+    1) Flatten the data structure
+    2) Rename certain fields to achieve consistency across data sets
+    3) Add additional fields to enhance usage of the data set in Socrata
+    (e.g. randomNum, coreData_position)
+
+	Parameters:
+		raw_rec: dictionary object of a single BSM record
+
+	Returns:
+		transformed dictionary object of the BSM record
+    '''
+    rename_prefix_fields = [
+        ('payload_data_coreData_', 'coreData_'),
+        ('coreData_accelSet_', 'coreData_accelset_')
+    ]
+    rename_fields = [
+        ('metadata_dataType', 'dataType'),
+        ('metadata_recordGeneratedAt', 'metadata_generatedAt'),
+        ('metadata_recordGeneratedBy', 'metadata_generatedBy'),
+        ('coreData_lat', 'coreData_position_lat'),
+        ('coreData_long', 'coreData_position_long'),
+        ('coreData_elev', 'coreData_elevation'),
+        ('coreData_accelset_yaw', 'coreData_accelset_accelYaw')
```

Comment:
What about the other accelset elements: pitch, and roll?

Reply:
The accelset element in BSM core uses "AccelerationSet4Way" and only contains long, lat, vert, and yaw, according to the J2735_201603.pdf document. Let me know if I should be referencing another schema standard document.

Reply:
Correct, but `coreData_position_lat`, `coreData_position_long` and `coreData_elevation' are not part of `coreData_accelset`. I believe we need total of 4 elements:

1. coreData_position_lat
2. coreData_position_lon
3. coreData_position_elev
4. coreData_accelset_lat
5. coreData_accelset_lon
6. coreData_accelset_vert
7. coreData_accelset_yaw

Reply:
Adding notes from our side discussion:
We do have the corresponding elements for the 7 fields listed above in the Tampa BSM dataset. 

1. coreData_position_lat
2. coreData_position_long
3. coreData_elevation (named this way to be consistent with WYDOT BSM schema on DTG)
4. coreData_accelset_lat
5. coreData_accelset_long
6. coreData_accelset_vert
7. coreData_accelset_accelYaw (named this way to be consistent with WYDOT BSM schema on DTG)